# Python Workshop

## NumPy

### Motivation

NumPy (Numerical Python) aims to provide array objects that are 50 times faster than traditional Python lists. It allows you to perform **vector and matrix calculations** in Python, and because many underlying functions are actually written in C, you can experience speeds that are impossible to find in native Python.

Array objects in NumPy are called `ndarray`, and they provide many supporting functions that make manipulation incredibly easy.

Arrays are widely used in data science because speed and resource utilization are crucial.

### Array

In [2]:
import numpy as np # Now the NumPy package can be referred to as np instead of numpy

# Create 1D Array
a = np.array([0, 1, 2, 3, 4])
b = np.array((0, 1, 2, 3, 4))
c = np.arange(5) # numpy.arange([start, ]stop, [step, ]), start_default = 0，step_default = 1
d = np.linspace(0, 2*np.pi, 5) # [0,2pi]，5 elements
print(a)
print(b)
print(c)
print(d)
print()

# 2-D array
arr2 = np.array([[1, 2, 3], [4, 5, 6]])
print(arr2)


[0 1 2 3 4]
[0 1 2 3 4]
[0 1 2 3 4]
[0.         1.57079633 3.14159265 4.71238898 6.28318531]

[[1 2 3]
 [4 5 6]]


* type(): This built-in Python function tells us the type of the object passed to it. Like the code above, it shows that `a` is of type `numpy.ndarray`.

To create an `ndarray`, we can pass a list, tuple, or any array-like object to the `array()` method, which will be converted to an `ndarray`.

### Higher Dimensional Arrays
When creating an array, you can use the `ndmin` parameter to define the number of dimensions.

In [3]:
# Create an array with 5 dimensions
arr = np.array([1, 2, 3, 4], ndmin = 5)

print(arr)
print('number of dimensions: ', arr.ndim)

[[[[[1 2 3 4]]]]]
number of dimensions:  5


### NumPy Array Indexing
In NumPy arrays, indices start from 0, meaning the first element has an index of 0, the second element has an index of 1, and so on.

In [4]:
# Access 1-D array
arr = np.array([1, 2, 3, 4])
print(arr[0])
print(arr[1] + arr[2])

# Access 2-D Arrays [(row-1),(col-1)]
arr = np.array([[1,2,3,4,5], [6,7,8,9,10]])
print('2nd element on 1st row: ', arr[0, 1])
print('5th element on 2nd row: ', arr[1, 4])

1
5
2nd element on 1st row:  2
5th element on 2nd row:  10


### Negative Indexing
We use negative indexing to access the array from the end of the array.

In [5]:
arr = np.array([[1,2,3,4,5], [6,7,8,9,10]])

print('Last element from 2nd dim: ', arr[1, -1])

Last element from 2nd dim:  10


Some useful creation functions(Optional)

1. `numpy.eye()` is a function in the NumPy library used to create a two-dimensional array (matrix) with 1s on the specified diagonal and zeros at other positions.

In [6]:
print(np.eye(3)) #单位矩阵
print(np.eye(3, 5))
print(np.eye(3, 5, k = 0))
print(np.eye(3, 5, k = 1))

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
[[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]]
[[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]]
[[0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]]


2. `numpy.diag()` is a function in the NumPy library used to create or extract diagonal arrays.

In [7]:
print(np.diag([1, 2, 3]))#对角矩阵

print(np.diag([1, 2, 3], 1))

a = np.array([[1, 2], [3, 4]]) #返回对角的值
b = np.array([[1, 2, 3],[4, 5, 6]])
print(np.diag(a))
print(np.diag(b))

[[1 0 0]
 [0 2 0]
 [0 0 3]]
[[0 1 0 0]
 [0 0 2 0]
 [0 0 0 3]
 [0 0 0 0]]
[1 4]
[1 5]


3. `numpy.zeros()` initializes all elements to 0, and `numpy.ones()` initializes all elements to 1.

In [8]:
print(np.zeros((2,3)))
print(np.zeros((2, 3, 2)))
print(np.ones((2, 3)))
print(np.ones((2, 3, 2)))

[[0. 0. 0.]
 [0. 0. 0.]]
[[[0. 0.]
  [0. 0.]
  [0. 0.]]

 [[0. 0.]
  [0. 0.]
  [0. 0.]]]
[[1. 1. 1.]
 [1. 1. 1.]]
[[[1. 1.]
  [1. 1.]
  [1. 1.]]

 [[1. 1.]
  [1. 1.]
  [1. 1.]]]


4. Creating a Random Matrix(Optional)

The `numpy.random` module implements a **pseudo-random number** generator (PRNG, or RNG for short), capable of drawing samples from various probability distributions. Typically, users will create a `Generator` instance using `default_rng` and call its various methods to obtain samples from different distributions.

In [9]:
from numpy.random import default_rng
rng1 = np.random.default_rng(42)
print(rng1.random((2,3)))

rng2 = np.random.default_rng(42) 
print(rng2.random((2,3,2)))

# The seed value is 42 bits, and you can use `secrets.randbits(128)` to get any 128-bit integer as the seed value.


[[0.77395605 0.43887844 0.85859792]
 [0.69736803 0.09417735 0.97562235]]
[[[0.77395605 0.43887844]
  [0.85859792 0.69736803]
  [0.09417735 0.97562235]]

 [[0.7611397  0.78606431]
  [0.12811363 0.45038594]
  [0.37079802 0.92676499]]]


### Slicing

In Python, slicing means moving elements from one given index to another.

Slicing Syntax

  ```python
  print(array[start:end])

  print(array[start:end:step])
  ```

`start` defaults to 0, `end` defaults to the length of the array, and `step` defaults to 1. 

**Caution:** `start` is included, but `end` is not.

In [10]:
arr = np.array([1, 2, 3, 4, 5, 6, 7])

print(arr[1:5])

print(arr[-3:]) # [end-3, end]

print(arr[-3:-1]) # array[end-3, end-1]

print(arr[1:5:2]) # Caution: [index1,index5), step = 2


[2 3 4 5]
[5 6 7]
[5 6]
[2 4]


  **Quiz**: 输出二维数组的
1. 奇数行偶数列元素 Odd row even column elements
2. 第二行 Second row
3. 第三列 Third column
4. 第一行的第二第四个元素 The second and fourth elements in the first row
5. 第二第四行的第一第三个元素(共四个数) The first and third elements of the second and fourth rows (a total of four numbers)

  **Answer(Complete the blanks)**

In [ ]:
a = np.array(
    [[11, 12, 13, 14, 15],
    [16, 17, 18, 19, 20],
    [21, 22, 23, 24, 25],
    [26, 27, 28 ,29, 30],
    [31, 32, 33, 34, 35]])

print()

print()

print()

print()



[[12 14]
 [22 24]
 [32 34]]

[16 17 18 19 20]

[13 18 23 28 33]

[12 14]

[16 18 26 28]


### Array Properties (Optional)

In [12]:
a = np.array([
            [11, 12, 13, 14, 15],
            [16, 17, 18, 19, 20],
            [21, 22, 23, 24, 25],
            [26, 27, 28 ,29, 30],
            [31, 32, 33, 34, 35]])

print(type(a)) # >>><class 'numpy.ndarray'>
print(a.dtype) # >>>int64
print(a.size) # >>>25
print(a.shape) # >>>(5, 5)
print(a.itemsize) # >>>8
print(a.ndim) # >>>2
print(a.nbytes) # >>>200


<class 'numpy.ndarray'>
int64
25
(5, 5)
8
2
200


In [13]:
# Creating Arrays With a Defined Data Type
arr = np.array([1, 2, 3, 4], dtype='S')
print()
print(arr)
print(arr.dtype)
print()

# Converting Data Type on Existing Arrays
arr = np.array([1.1, 0, 3])
newarr = arr.astype(bool)
print(newarr)
print(newarr.dtype)


[b'1' b'2' b'3' b'4']
|S1

[ True False  True]
bool


If elements of a given type cannot be type-casted, NumPy will throw a **ValueError** exception.

- `a.shape`

The shape of an array is the number of its rows and columns. The array above has 5 rows and 5 columns, so its shape is (5, 5).

- `a.itemsize`

The `itemsize` property is the number of bytes occupied by each item. This array's data type is `int 64`. An `int 64` has 64 bits, and a byte has 8 bits. Dividing by 64 and then by 8 gives you the number of bytes occupied; in this example, it's 8.

- `a.ndim`

The `ndim` property is the dimension of the array. This one has 2. For example, a vector has only 1.

- `a.nbytes`

The `nbytes` property is the number of bytes consumed by all the data in the array. You should note that this does not take into account the array's overhead, so the actual space occupied by the array will be slightly larger.

### Calculation

The term "broadcast" describes how NumPy handles arrays of different shapes during arithmetic operations. With certain constraints, smaller arrays are "broadcast" to larger arrays so that they have compatible shapes. 

 However, in some cases, broadcasting is not a good idea because it can lead to inefficient memory usage, thus slowing down computation.

In [14]:
a = np.array([1.0, 2.0, 3.0])
b = np.array([2.0, 2.0, 2.0])
print(a * b)

a = np.array([1.0, 2.0, 3.0])
b = 2.0
print(a * b)

print(a+b)

[2. 4. 6.]
[2. 4. 6.]
[3. 4. 5.]


In [15]:
from numpy import array, argmin, sqrt, sum
observation = array([111.0, 188.0])
codes = array([
               [102.0, 203.0],
               [132.0, 193.0],
               [45.0, 155.0],
               [57.0, 173.0]])
diff = codes - observation    # the broadcast happens here
print(diff)

[[ -9.  15.]
 [ 21.   5.]
 [-66. -33.]
 [-54. -15.]]


 For a more in-depth understanding [see](https://numpy.org/doc/stable/user/basics.broadcasting.html#broadcastable-arrays)

  - 点乘与叉乘

In [16]:
# dot product
a = np.array([1,0,0])
b = np.array([0,1,0])
print(np.dot(a,b))

0


In [17]:
# cross product
a = np.array([2,0,0])
b = np.array([2,2,0])
print(np.cross(a,b)) 

a = np.array([1,0,0])
b = np.array([0,1,0])
print(np.cross(a,b))
print(np.cross(b,a))

[0 0 4]
[0 0 1]
[ 0  0 -1]


- 逆矩阵和转置

In [18]:
# Get Inverse and Transpose

a=np.matrix([[1, 2, 3],[4, 5, 6]])

print(a)

print(a.I)

print(a.T)

[[1 2 3]
 [4 5 6]]
[[-0.94444444  0.44444444]
 [-0.11111111  0.11111111]
 [ 0.72222222 -0.22222222]]
[[1 4]
 [2 5]
 [3 6]]


[see](https://numpy.org/doc/stable/reference/generated/numpy.matrix.html)

- 求和，最大最小值，累积求和（从前往后不断加）  
Summation, maximum and minimum values, cumulative summation (adding continuously from front to back).

In [19]:
a = np.arange(10)
print(a)
print(a.sum()) # >>>45
print(a.min()) # >>>0
print(a.max()) # >>>9
print(a.cumsum()) # >>>[ 0  1  3  6 10 15 21 28 36 45]

[0 1 2 3 4 5 6 7 8 9]
45
0
9
[ 0  1  3  6 10 15 21 28 36 45]


Linear Algebra(Optional)

In [20]:
from numpy import linalg as LA
matrix = np.diag((1, 2, 3))
print(matrix)
print()
eigenvalues, eigenvectors = LA.eig(matrix)
print(eigenvalues) # 特征值
print(eigenvectors) # 特征向量矩阵(按'列'存储)

[[1 0 0]
 [0 2 0]
 [0 0 3]]

[1. 2. 3.]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


[参见](https://numpy.org/doc/stable/reference/routines.linalg.html)

### Copying, Splicing, or Modifying Existing Arrays

1. Copying `view` and `copy`

View and copy can be compared to referencing and copying. Modifications to a view directly affect the original data, while modifications to a copy do not.

- Copy owns the data; any changes made to a copy will not affect the original array, and vice versa.

In [21]:
arr = np.array([1, 2, 3, 4, 5])
x = arr.copy()
arr[0] = 42

print(arr)
print(x)

[42  2  3  4  5]
[1 2 3 4 5]


- view  
The view does not own the data; any modification made to the view will affect the original array, and any modification made to the original array will also affect the view. (Similar to slicing)


In [22]:
arr = np.array([1, 2, 3, 4, 5])
x = arr.view()
arr[0] = 42

print(arr)
print(x)

[42  2  3  4  5]
[42  2  3  4  5]


In [23]:
a = np.array([1, 2, 3, 4, 5, 6])
b = a[:2]
b +=1
print('a =', a, '; b =', b)

a = [2 3 3 4 5 6] ; b = [2 3]


How to determine if it's a copy:

- Empirical method: Low-dimensional indexes (contiguous memory) usually produce views, while high-dimensional indexes (non-contiguous memory) usually produce copies.

- Code testing:

**quiz**: What is the output of the following code?
``` python
a = np.arange(10)
b = a[2:6]
b = b * 2
print(a)
```
A. [2 3 4 5]
B. [0 1 2 3 4 5 6 7 8 9]
C. [ 4  6  8 10]
D. [ 0  1  2  3  4  5  6  7  8  9 10]


1. matrix splicing

In [24]:
A = np.ones((2, 2))
B = np.eye(2, 2)
C = np.zeros((2, 2))
D = np.diag((-3, -4))
np.block([[A, B], [C, D]])

array([[ 1.,  1.,  1.,  0.],
       [ 1.,  1.,  0.,  1.],
       [ 0.,  0., -3.,  0.],
       [ 0.,  0.,  0., -4.]])

3. Reshape

    Reshape a matrix while keeping its contents unchanged, but require that the number of elements in the matrix before and after reshape is the same; otherwise, an error will occur!

In [25]:
arr = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12])

newarr = arr.reshape(4, 3) # change to a 4*3 matrix

print(newarr)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


`reshape()` allows for an "unknown" dimension. This means that in the `reshape` method, you don't have to specify an exact value for one of the dimensions. Passing -1 as the value will allow NumPy to calculate that number for you.

In [26]:
arr = np.array([1, 2, 3, 4, 5, 6, 7, 8])

newarr = arr.reshape(2, 2, -1)

print(newarr)

[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


**Quiz**: 将矩阵[1,2,3]改写成[[1,2,3],[4,5,6],[7,8,9]]再将矩阵转化为一维数组。使用broadcast,拼接,reshape等技巧。  
Rewrite the matrix [1,2,3] as [[1,2,3],[4,5,6],[7,8,9]] and then convert the matrix into a one-dimensional array. Use techniques such as broadcasting, concatenation, and reshaping.  
**Answer**:

In [ ]:
# opt 1:


# opt 2 (use broadcast!):



[[1 2 3]
 [4 5 6]
 [7 8 9]]
[1 2 3 4 5 6 7 8 9]

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[1 2 3 4 5 6 7 8 9]


4. 布尔索引（Boolean Indexing）  

    Using True/False as "filters"

- If the condition is met → True → Selected

- If the condition is not met → False → Filtered

In [28]:
a = np.array([1,2,3,4,5,6]) # Masking
print(a[a<4])

[1 2 3]


In [29]:
a = np.arange(0, 100, 10)
b = a[:5] # from 0th element to 4th element
c = a[a >= 50]

print(a) # >>>[ 0 10 20 30 40 50 60 70 80 90]
print(b) # >>>[ 0 10 20 30 40]
print(c) # >>>[50 60 70 80 90]

[ 0 10 20 30 40 50 60 70 80 90]
[ 0 10 20 30 40]
[50 60 70 80 90]


In [30]:
# Where：返回值是满足条件的索引位置,而且返回值是一个 tuple!!(非常关键)
a = np.arange(0, 100, 10)
b = np.where(a < 50) 
c = np.where(a >= 50)[0]
print(a)
print(b) # >>>(array([0, 1, 2, 3, 4]),) 是一个长度为1的tuple
print(c) # >>>[5 6 7 8 9] 是ndarray

[ 0 10 20 30 40 50 60 70 80 90]
(array([0, 1, 2, 3, 4]),)
[5 6 7 8 9]


### Vectorization

`vectorize` transforms a function that can only process a single element into a function that can be applied to an array.   
  `vectorize` provides this functionality primarily for convenience, not performance. The implementation is essentially a for loop.

In [31]:
def myfunc(a, b):
    "Return a-b if a>b, otherwise return a+b"
    if a > b:
        return a - b
    else:
        return a + b

vfunc = np.vectorize(myfunc)
print(vfunc([1, 2, 3, 4], 2))

# equivalence code
a = [1,2,3,4]
print(np.array([myfunc(x,2) for x in a]))


[3 4 1 2]
[3 4 1 2]


Answer of the quiz:  
1.**quiz**: What is the output of the following code?
``` python
a = np.arange(10)
b = a[2:6]
b = b * 2
print(a)
```
A. [2 3 4 5]  
**B. [0 1 2 3 4 5 6 7 8 9]**  
C. [ 4  6  8 10]  
D. [ 0  1  2  3  4  5  6  7  8  9 10]


## Pandas
`Pandas` enables us to process large amounts of data, and is especially well-suited for handling structured data (such as tables, time series, etc.) and drawing conclusions based on statistical principles.      
`Pandas` can clean up messy datasets, making them easy to read and relevant.


### Series 
A Series is a one-dimensional array that can store data of any type.

Labels:   
Unless otherwise specified, values ​​are labeled using their index numbers. The first value is indexed as 0, the second as 1, and so on.

In [32]:
# Create series
import pandas as pd

a = [1, 7, 2]
myvar = pd.Series(a) 

print(myvar)
print(myvar[0])

0    1
1    7
2    2
dtype: int64
1


You can also customize the label name using the `index` parameter (e.g., "x", "y", "z", etc.).

In [33]:
a = [1, 7, 2]
myvar = pd.Series(a, index = ["x", "y", "z"])

print(myvar)
print(myvar["y"])

x    1
y    7
z    2
dtype: int64
7


When creating a Series, you can also use key-value pair objects (dictionaries).

**Note:** The keys of the dictionary will become the tags.

In [34]:
calories = {"day1": 420, "day2": 380, "day3": 390}

# select only some of the items by using index argument
myvar = pd.Series(calories, index = ["day1", "day2"])

print(myvar)

day1    420
day2    380
dtype: int64


### DataFrames  
A Pandas DataFrame is a two-dimensional data structure, similar to a two-dimensional array or a table with rows and columns.

**Note:** A Series is like a column; a DataFrame is the entire table.

1. Creating a DataFrame

In [35]:
data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

print(df) 

   calories  duration
0       420        50
1       380        40
2       390        45


2. Locating rows/columns:  
    Pandas uses the `loc` attribute to return one or more specified rows.

In [36]:
# Return row 0, refer to the row index
print(df.loc[0])
print()
print(df.loc[[0,1]])

calories    420
duration     50
Name: 0, dtype: int64

   calories  duration
0       420        50
1       380        40


3. Named Index

In [38]:
newdf = pd.DataFrame(data, index = ["day1", "day2", "day3"])
print(newdf)
print()
#refer to the named index:
print(newdf.loc[["day2","day1"]]) # use label
print(newdf.iloc[[1,0]]) # use index

      calories  duration
day1       420        50
day2       380        40
day3       390        45

      calories  duration
day2       380        40
day1       420        50
      calories  duration
day2       380        40
day1       420        50


4. Data load    
(1) CSV (Comma Separated Values)

In [41]:
df = pd.read_csv('data.csv')

print(df)

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

In [611]:
print(df.head(3)) # print the first 10 rows
print(df.tail()) # print the last 5 rows

   Duration          Date  Pulse  Maxpulse  Calories
0        60  '2020/12/01'    110       130     409.1
1        60  '2020/12/02'    117       145     479.0
2        60  '2020/12/03'    103       135     340.0
    Duration          Date  Pulse  Maxpulse  Calories
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0


(2) JSON  
Large datasets are typically stored or retrieved in JSON format.

JSON is a plain text format, but it has an object structure.

In [612]:
df = pd.read_json('data.json')
print(df) 

     Duration  Pulse  Maxpulse  Calories
0          60    110       130     409.1
1          60    117       145     479.0
2          60    103       135     340.0
3          45    109       175     282.4
4          45    117       148     406.0
..        ...    ...       ...       ...
164        60    105       140     290.8
165        60    110       145     300.4
166        60    115       145     310.2
167        75    120       150     320.4
168        75    125       150     330.4

[169 rows x 4 columns]


### Data Cleaning

Data cleaning refers to fixing erroneous data in a dataset.

Erroneous data may include:

- Empty cells

- Incorrect data format

- Data errors

- Duplicate data

1. **Empty Cells**

(1) Delete rows containing empty cells

By default, the `df.dropna()` method returns a new DataFrame without modifying the original DataFrame.

If you need to modify the original DataFrame, use `df.dropna(inplace = True)`.

In [613]:
df = pd.read_csv('data.csv')
print(df.iloc[[18,22,28]])
new_df = df.dropna()
print(new_df.iloc[[18,22,28]])

    Duration          Date  Pulse  Maxpulse  Calories
18        45  '2020/12/18'     90       112       NaN
22        45           NaN    100       119     282.0
28        60  '2020/12/28'    103       132       NaN
    Duration          Date  Pulse  Maxpulse  Calories
19        60  '2020/12/19'    103       123     323.0
24        45  '2020/12/24'    105       132     246.0
31        60  '2020/12/31'     92       115     243.0


(2) Replacing empty values  

Another way to handle empty cells is to insert new values.

In [614]:
# Replace all empty cells
df = pd.read_csv('data.csv')
df.fillna(130, inplace = True)
print(df.iloc[[18,22,28]])

    Duration          Date  Pulse  Maxpulse  Calories
18        45  '2020/12/18'     90       112     130.0
22        45           130    100       119     282.0
28        60  '2020/12/28'    103       132     130.0


In [ ]:
# Replace only for specified columns
df.fillna({"Calories": 130}, inplace = True)
print(df.iloc[[18,22,28]])

# Replace using mean
df = pd.read_csv("data.csv")
x = df["Calories"].mean()
df.fillna({"Calories": x}, inplace = True)

# Replace using median
x = df["Calories"].median()
df.fillna({"Calories": x}, inplace = True)

# Replace using mode (the value that appears most frequently)
x = df["Calories"].mode()[0]
df.fillna({"Calories": x}, inplace = True)

    Duration          Date  Pulse  Maxpulse  Calories
18        45  '2020/12/18'     90       112     130.0
22        45           NaN    100       119     282.0
28        60  '2020/12/28'    103       132     130.0


2. **Incorrect Data Format**

(1) Convert to the correct format

In [616]:
df = pd.read_csv('data.csv')
print(df.iloc[[21,22,26]])

df['Date'] = pd.to_datetime(df['Date'], format='mixed')
print(df.iloc[[21,22,26]])

    Duration          Date  Pulse  Maxpulse  Calories
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
26        60    2020/12/26    100       120     250.0
    Duration       Date  Pulse  Maxpulse  Calories
21        60 2020-12-21    108       131     364.2
22        45        NaT    100       119     282.0
26        60 2020-12-26    100       120     250.0


(2) Delete the entire line

**Quiz**: Remember the command we have learned?  
df._______(subset=['Date'], inplace = True)

In [617]:
df.dropna(subset=['Date'], inplace = True)
print(df.iloc[[21,22,23]])

    Duration       Date  Pulse  Maxpulse  Calories
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0


3. Incorrect Data

e.g., the duration in row 7 is 450, while the durations in all other rows are between 30 and 60.

This isn't necessarily incorrect, but considering this is a dataset of someone's workout logs, we can conclude that this person's actual workout time did not reach 450 minutes.

In [618]:
# (1) Replacing Values
print(df.iloc[[6,7]])
df.loc[7, 'Duration'] = 45
print(df.iloc[[6,7]])
# For wrong data for larger data sets
for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.loc[x, "Duration"] = 120

# (2) Removing Rows
for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.drop(x, inplace = True)

   Duration       Date  Pulse  Maxpulse  Calories
6        60 2020-12-07    110       136     374.0
7       450 2020-12-08    104       134     253.3
   Duration       Date  Pulse  Maxpulse  Calories
6        60 2020-12-07    110       136     374.0
7        45 2020-12-08    104       134     253.3


4. Duplicate Items

(1) Discovering Duplicate Items

In [619]:
print(df.iloc[[11,12]])
print()
print(df.duplicated())

    Duration       Date  Pulse  Maxpulse  Calories
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
dtype: bool


(2) Remove duplicate items

In [620]:
df.drop_duplicates(inplace = True)
print(df.iloc[[11,12]])

    Duration       Date  Pulse  Maxpulse  Calories
11        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3


### Data Correlation

The `corr()` method is used to calculate the relationship between each column in a dataset.

**Note:** The `corr()` method ignores "not numeric" columns.

In [621]:
df.corr()

,Duration,Date,Pulse,Maxpulse,Calories
Duration,1.000000,0.212084,-0.083417,-0.296585,0.350969
Date,0.212084,1.000000,-0.369328,-0.517827,-0.368101
Pulse,-0.083417,-0.369328,1.000000,0.261426,0.503243
Maxpulse,-0.296585,-0.517827,0.261426,1.000000,0.338515
Calories,0.350969,-0.368101,0.503243,0.338515,1.000000


How should the results be interpreted?

- Values ​​range from -1 to 1.

- 1 indicates a one-to-one relationship (perfect correlation). For this dataset, an increase in one value in the first column will also increase the value in the other column.

- 0.9 also represents a good correlation; increasing one value will likely increase the value in the other column.

- -0.9 has a similarly good correlation to 0.9, but increasing one value will likely decrease the value in the other column.

- 0.2 indicates a very poor correlation, meaning that an increase in one value does not necessarily mean that an increase in the other value will also occur.

### View other attributes (Optional)

In [622]:
print(pd.__version__) # Checking Pandas Version
print()
print(df.info()) # Print information about the data

2.3.3

<class 'pandas.core.frame.DataFrame'>
Index: 30 entries, 0 to 31
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   Duration  30 non-null     int64         
 1   Date      30 non-null     datetime64[ns]
 2   Pulse     30 non-null     int64         
 3   Maxpulse  30 non-null     int64         
 4   Calories  28 non-null     float64       
dtypes: datetime64[ns](1), float64(1), int64(3)
memory usage: 1.4 KB
None


### Reference
[2023 zzjc's part](https://github.com/UMJI-SSTIA/Python-wksp)

useful self-study website:  
[numpy](https://www.w3schools.com/python/numpy/numpy_array_reshape.asp)  
[panda](https://www.w3schools.com/python/pandas/default.asp)